cmake/Corrosion.cmake

```diff
@@ -1397,3 +1397,74 @@ function(corrosion_install)
         message(FATAL_ERROR "install(EXPORT ...) not yet implemented")
     endif()
 endfunction()
+
+# Create a library target that will run cxxbridge on the given files on a crate
+# that is assumed to contain a module with #[cxx::bridge]
+# Call with corrosion_cxxbridge(<target> CRATE <crate> PATH <tomlPath> FILES [<filePath>, ...])
+# <target> is the name of the target created that will host the cxx interface
+# CRATE <crate> is the target name of the specific target generated by corrosion_import_crate
+# PATH <tomlPath> is the relative path from where this function is called to Cargo.toml of <crate>
+# FILES [<filePath, ...] is the relative paths to files with #[cxx::bridge]. These are relative to <tomlPath>
+# 
+# Example: corrosion_cxxbridge(myCxxTarget CRATE someCrate PATH someCratePath FILES src/lib.rs)
+function(corrosion_cxxbridge target)
+    # This function is based on the details given by https://cxx.rs/build/other.html#producing-the-generated-code
+    # corrosion_cxxbridge requires the cxxbridge executable to be installed
+    # run `cargo install cxxbridge-cmd` if missing
+    find_program(CXXBRIDGE cxxbridge PATHS "$ENV{HOME}/.cargo/bin/" REQUIRED)
+
+    set(OPTIONS)
+    set(ONE_VALUE_KEYWORDS CRATE PATH)
+    set(MULTI_VALUE_KEYWORDS FILES)
+    cmake_parse_arguments(_arg "${OPTIONS}" "${ONE_VALUE_KEYWORDS}" "${MULTI_VALUE_KEYWORDS}" ${ARGN})
+
+    if("${_arg_CRATE}" STREQUAL "")
+        message(FATAL_ERROR "corrosion_cxxbridge called without a crate")
+    endif()
+
+    if("${_arg_PATH}" STREQUAL "")
+        message(FATAL_ERROR "corrosion_cxxbridge called without path to folder of Cargo.toml of target ${_arg_CRATE}")
+    endif()
+
+    if(NOT EXISTS "${CMAKE_CURRENT_LIST_DIR}/${_arg_PATH}/Cargo.toml")
+        message(FATAL_ERROR "The path ${CMAKE_CURRENT_LIST_DIR}/${_arg_PATH} does not contain a Cargo.toml")
+    endif()
+
+    if("${_arg_FILES}" STREQUAL "")
+        message(FATAL_ERROR "Must supply at least the path of a single file to be bridged, defined as relative path from Cargo.toml")
+    endif()
+
+    set(crate_path ${_arg_PATH})
+
+    add_library(${target})
+    target_include_directories(${target} PUBLIC ${CMAKE_CURRENT_BINARY_DIR}/cxxgenerated)
+    target_link_libraries(${target} PRIVATE ${_arg_CRATE})
+
+    set(cxx_generated_dir ${CMAKE_CURRENT_BINARY_DIR}/cxxgenerated/${target})
+
+    foreach(file ${_arg_FILES})
+        get_filename_component(filename ${file} NAME_WE)
+        set(cxx_header ${filename}.h)
+        set(cxx_source ${filename}.cpp)
```

Comment:
For multiple files like the common `mod.rs` name but in different directories, we would generate just one source file, since we would overwrite it. I guess for a first version we could just leave it as is and document as a known limitation for now.
I'm not sure if we really want to preserve the directory structure though. At least the common `src` directory we would probably want to strip...

Reply:
Regarding the `src`, I thought it was helpful in the sense that they are all relative to each other. But src as far as I can see must always be included regardless, so maybe its not confusing to omit it.

Regarding if we have multiple levels of nesting, I could test out a change to see if I can match the folder structure as well!

Reply:
What we probably want to avoid is for C++ files having to include like this:
```C++
#include <Mytarget/src/myheader.h>
//                 ^^^ This would be strange
```

Maybe we could also change the directory structure slightly like this: 
```cmake
# Make a corrosion_generated directory in which also other artifacts, e.g. bindgen, cbindgen could be placed
set(corrosion_generated_dir ${CMAKE_CURRENT_BINARY_DIR}/corrosion_generated/${target})
# subdirectory for cxx artifacts
set(cxx_generated_dir ${corrosion_generated_dir }/cxx)
target_include_directories(${target} PUBLIC ${CMAKE_CURRENT_BINARY_DIR}/corrosion_generated/${target}/)
# Or maybe optionally allow to toggle the prefix and instead include like
target_include_directories(${target} PUBLIC ${CMAKE_CURRENT_BINARY_DIR}/corrosion_generated/${target}/cxx)
```

What do you think?

Reply:
Yeah the first thing there is definitely strange and is something that is already avoided with the current implementation.
It however places all files in the same directory as `#include "<cxxbridge_target_name>/<rust_source_filename_we>.h"`
where
`corrosion_cxxbridge(myCxxTarget CRATE someCrate PATH someCratePath FILES src/lib.rs src/foo/lib.rs)`
will both add a file placed as `myCxxTarget/lib.h`, which is suboptimal.

Regarding the second part, I am a bit unsure. Have you some typo regarding which variable is used?
But yes, standardizing on a corrosion_generated folder with subtypes for different generations is definitely smart, will absolutely do!

Reply:
Changed to that it now follows the folder structure

``` cmake
corrosion_cxxbridge(myCxxTarget CRATE someCrate PATH someCratePath FILES lib.rs foo/lib.rs)
```

will now result in

``` cpp
#include "myCxxTarget/lib.h"
#include "myCxxTarget/foo/lib.h"
```

Reply:
> Regarding the second part, I am a bit unsure. Have you some typo regarding which variable is used?

I edited my comment above. We could probably use CMake to strip the `src/` prefix if present.

> will now result in
> #include "myCxxTarget/lib.h"
> #include "myCxxTarget/foo/lib.h"

That sounds nice. I haven't had the time to look at that part again, I'll have another look on the weekend.

Reply:
Hmm, but it is stripped. You will never have to write src anywhere with the current solution. So it should be as you wanted it to be, unless I misunderstood again 😊